File: morphac/mechanics/models/test/dubin_model_test.cc
#include "mechanics/models/include/dubin_model.h"

#include "Eigen/Dense"
#include "gtest/gtest.h"

namespace {

using Eigen::VectorXd;

using morphac::constructs::ControlInput;
using morphac::constructs::State;
using morphac::mechanics::models::DubinModel;
using morphac::robot::blueprint::Footprint;

class DubinModelTest : public ::testing::Test {
 protected:
  DubinModelTest() {
    // Set random seed for Eigen.
    srand(7);
  }

  void SetUp() override{};
};

TEST_F(DubinModelTest, Construction) {
  DubinModel dubin_model{1.5};

  ASSERT_EQ(dubin_model.pose_size, 3);
  ASSERT_EQ(dubin_model.velocity_size, 0);
  ASSERT_EQ(dubin_model.control_input_size, 1);
  ASSERT_EQ(dubin_model.speed, 1.5);
}

TEST_F(DubinModelTest, DerivativeComputation) {
  DubinModel dubin_model{2.5};
  VectorXd pose_vector1(3), pose_vector2(3);
  VectorXd control_input_vector1(1), control_input_vector2(1);
  VectorXd desired_vector1(3), desired_vector2(3);
  pose_vector1 << 9, 10, M_PI / 2;
  pose_vector2 << 1.5, 2.3, 0.;
  control_input_vector1 << 5.5;
  control_input_vector2 << -1.5;
  desired_vector1 << 0., 2.5, 5.5;
  desired_vector2 << 2.5, 0., -1.5;

  State state1{pose_vector1, VectorXd::Zero(0)};
  ControlInput control_input1{control_input_vector1};

  // Computing and verifying the derivative computation for different control
  // inputs.
  State derivative1 =
      dubin_model.ComputeStateDerivative(state1, control_input1);
  ASSERT_TRUE(derivative1.get_pose_data().isApprox(desired_vector1));
  ASSERT_TRUE(derivative1.IsVelocityEmpty());

  State derivative2 =
      dubin_model.ComputeStateDerivative(State{pose_vector2, VectorXd::Zero(0)},
                                         ControlInput{control_input_vector2});
  ASSERT_TRUE(derivative2.get_pose_data().isApprox(desired_vector2));
  ASSERT_TRUE(derivative2.IsVelocityEmpty());
}

TEST_F(DubinModelTest, InvalidDerivativeComputation) {
  DubinModel dubin_model{1.0};

  // Computing the state derivative with incorrect state/control
  // input/derivative.
  ASSERT_THROW(dubin_model.ComputeStateDerivative(State(2, 0), ControlInput(1)),
               std::invalid_argument);
  ASSERT_THROW(dubin_model.ComputeStateDerivative(State(4, 0), ControlInput(1)),
               std::invalid_argument);
  ASSERT_THROW(dubin_model.ComputeStateDerivative(State(3, 1), ControlInput(1)),
               std::invalid_argument);
  ASSERT_THROW(dubin_model.ComputeStateDerivative(State(3, 0), ControlInput(0)),
               std::invalid_argument);
  ASSERT_THROW(dubin_model.ComputeStateDerivative(State(3, 0), ControlInput(2)),
               std::invalid_argument);
}

TEST_F(DubinModelTest, StateNormalization) {
  DubinModel dubin_model{1};

  // Making sure that the angle gets normalized.
  State state1({0, 0, 2 * M_PI}, {});
  State state2({0, 0, 2 * M_PI + 4 * M_PI / 3.}, {});
  State state3({0, 0, -2 * M_PI - 4 * M_PI / 3.}, {});
  State normalized_state1({0, 0, 0}, {});
  State normalized_state2({0, 0, -2 * M_PI / 3.}, {});
  State normalized_state3({0, 0, 2 * M_PI / 3.}, {});

  ASSERT_TRUE(dubin_model.NormalizeState(state1) == normalized_state1);
  ASSERT_TRUE(dubin_model.NormalizeState(state2) == normalized_state2);
  ASSERT_TRUE(dubin_model.NormalizeState(state3) == normalized_state3);
}

TEST_F(DubinModelTest, DefaultFootprint) {
  DubinModel dubin_model{1};

  Footprint footprint = dubin_model.DefaultFootprint();

  // Just make sure that the footprint is a triangle.
  ASSERT_EQ(footprint.get_data().rows(), 3.);
}

}  // namespace

int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: morphac/mechanics/models/python/test/dubin_model_test.py
import numpy as np
import pytest

from morphac.constructs import ControlInput, State
from morphac.mechanics.models import DubinModel
from morphac.robot.blueprint import Footprint


@pytest.fixture()
def generate_dubin_model_list():

    d1 = DubinModel(1)
    d2 = DubinModel(speed=2.5)

    return d1, d2


def test_speed(generate_dubin_model_list):

    d1, d2 = generate_dubin_model_list

    assert d1.speed == 1.0
    assert d2.speed == 2.5

    # Make sure that speed is read only.
    with pytest.raises(AttributeError):
        d1.speed = 2.0
    with pytest.raises(AttributeError):
        d2.speed = 3.0


def test_size(generate_dubin_model_list):

    d1, d2 = generate_dubin_model_list

    # The sizes are fixed.
    assert d1.pose_size == 3
    assert d2.pose_size == 3

    assert d1.velocity_size == 0
    assert d2.velocity_size == 0

    assert d1.control_input_size == 1
    assert d2.control_input_size == 1

    # Make sure that the sizes are read only.
    with pytest.raises(AttributeError):
        d1.pose_size = 2
    with pytest.raises(AttributeError):
        d2.velocity_size = 0
    with pytest.raises(AttributeError):
        d1.control_input_size = 3


def test_derivative_computation(generate_dubin_model_list):

    d1, d2 = generate_dubin_model_list

    # We only test that the function is called and returns properly, not the
    # actual computation as that cpp side tests this.

    # Test with positional arguments.
    der1 = d1.compute_state_derivative(
        robot_state=State([1, 2, 0], []), control_input=ControlInput(1)
    )

    # In this case, the control input controls the angular velocity and the
    # speed i# constant. Hence the robot must move horizontally (theta=0) with
    # the given speed.
    assert np.allclose(der1.data, [1, 0, 0])

    # Making sure that the derivative can only be computed from control inputs
    # that are of the correct dimensions.
    with pytest.raises(ValueError):
        d1.compute_state_derivative(State(0, 3), ControlInput(1))
    with pytest.raises(ValueError):
        d1.compute_state_derivative(State(3, 1), ControlInput(1))
    with pytest.raises(ValueError):
        d2.compute_state_derivative(State(2, 0), ControlInput(1))
    with pytest.raises(ValueError):
        d2.compute_state_derivative(State(3, 0), ControlInput(2))


def test_normalize_state():

    dubin_model = DubinModel(1)

    # As the cpp side tests the actual computation, we just check that the
    # normalize_state interface works.

    assert dubin_model.normalize_state(robot_state=State(3, 0)) == State(3, 0)


def test_default_footprint(generate_dubin_model_list):

    d1, _ = generate_dubin_model_list

    # Just making sure that the default_footprint function returns a valid footprint.
    assert isinstance(d1.default_footprint(), Footprint)
